### PR TITLE
Remove deprecated FITS code

### PR DESCRIPTION
**Pull request by saimn**
Following comments in #5006, I started to remove deprecated code in `io.fits`. The removed code was deprecated in Astropy 0.3/0.4, see also the list in the [changelog](http://docs.astropy.org/en/latest/changelog.html#id85). 
I updated several tests that were testing the old API, to use the new one, as I wasn't sure if these tests have an equivalent for the new API. 
cc @embray

### CODE REVIEW

File: astropy/io/fits/tests/test_header.py [hunk 1]
```diff
@@ -50,23 +50,7 @@ class TestOldApiHeaderFunctions(FitsTestCase):
...
-        """Test backwards compatibility support for Header.rename_key()"""
-        header = fits.Header([('A', 'B', 'C'), ('D', 'E', 'F')])
-        header.rename_key('A', 'B')
-        assert 'A' not in header
-        assert 'B' in header
-        assert header[0] == 'B'
-        assert header['B'] == 'B'
-        assert header.comments['B'] == 'C'
-
+    # FIXME: This one still works and does not trigger any warning
```
Comment (saimn) [inline astropy/io/fits/tests/test_header.py]: This test is the last one in the `TestOldApiHeaderFunctions` class. It does not trigger deprecation warning, and seems to be the only test for `add_history`. So I think I will move it to the `TestHeaderFunctions` class.

File: astropy/io/fits/tests/test_header.py [hunk 2]
```diff
@@ -715,54 +553,54 @@ def test_hierarch_create_and_update(self):
```
Comment (saimn) [inline astropy/io/fits/tests/test_header.py]: About the number of warnings that changed in these tests: it seems that the old update behavior was not triggering "HIERARCH" warnings, though the new behavior does.

File: astropy/io/fits/convenience.py [hunk 3]
```diff
@@ -725,7 +724,7 @@ def tabledump(filename, datafile=None, cdfile=None, hfile=None, ext=1,
     -----
     The primary use for the `tabledump` function is to allow editing in a
     standard text editor of the table data and parameters.  The
-    ``tcreate`` function can be used to reassemble the table from the
+    ``tableload`` function can be used to reassemble the table from the
```
Comment (MSeifert04) [inline astropy/io/fits/convenience.py]: any reason this is rendered as code instead of a link?

File: astropy/io/fits/hdu/compressed.py [hunk 4]
```diff
@@ -1362,21 +1362,6 @@ def _update_header_data(self, image_header,
             for _ in range(required_blanks - table_blanks):
                 self._header.append()
 
-    @deprecated('0.3', alternative='(refactor your code)', pending=True)
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/compressed.py]: Sad to lose this `alternative` :smile:

File: astropy/io/fits/tests/test_header.py [hunk 5]
```diff
@@ -715,54 +553,54 @@ def test_hierarch_create_and_update(self):
 
         header = fits.Header()
         with catch_warnings(VerifyWarning) as w:
-            header.update('HIERARCH BLAH BLAH', 'TESTA')
+            header.update({'HIERARCH BLAH BLAH': 'TESTA'})
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_header.py]: Is this necessary? I haven't looked at the code but is the removed command invalid now?

File: astropy/io/fits/convenience.py (hunk 3 above)
Comment (embray) [inline astropy/io/fits/convenience.py]: No reason I can think of, unless it's one of those odd cases where the reference isn't resolved for some reason.

**Comment by MSeifert04** [comment]
One further point would be to check the documentation if there are unresolved links or (unexecuted) ``codes`` that use any deprecated API functions.

File: astropy/io/fits/hdu/compressed.py (hunk 4 above)
Comment (embray) [inline astropy/io/fits/hdu/compressed.py]: I think what this really was, is that this method never should have been public in the first place and there's no reason anyone should have been using it directly.  But the message probably could have been clearer about that :)

Fortunately I don't think anyone actually *was* using this method.

File: astropy/io/fits/tests/test_header.py (hunk 1 above)
Comment (embray) [inline astropy/io/fits/tests/test_header.py]: Sounds good.

File: astropy/io/fits/tests/test_header.py [hunk 6]
```diff
@@ -99,148 +83,6 @@ def test_add_commentary(self):
...
-        assert cards[0].value == 0
-        assert cards[1].value == 2
-
-    @ignore_warnings(AstropyDeprecationWarning)
-    def test_assign_boolean(self):
-        """
-        Regression test for https://aeon.stsci.edu/ssb/trac/pyfits/ticket/123
-
-        Tests assigning Python and Numpy boolean values to keyword values.
-        """
```
Comment (embray) [inline astropy/io/fits/tests/test_header.py]: I'm hesitant about outright removing this test and some of the ones above.  Especially if they're regression tests.  Instead they should be rewritten to just not use the deprecated APIs.  I have a vague memory of keeping some tests using deprecated APIs just as a way to exercise those APIs.  But it doesn't mean the test itself isn't still relevant.

At the very least I'd want to know what coverage is changed by removing these tests.

File: astropy/io/fits/tests/test_header.py (hunk 5 above)
Comment (embray) [inline astropy/io/fits/tests/test_header.py]: Yes, change is correct.  `header.update` would no longer work as an alias for `header.set` with the deprecated code removed.

**Comment by embray** [comment]
Thanks for doing this. I'm sort of amazed by how much of this stuff was floating around.  I think I've always been hesitant to remove them because who knows what it will break at STScI.

File: astropy/io/fits/convenience.py (hunk 3 above)
Comment (saimn) [inline astropy/io/fits/convenience.py]: Done.

File: astropy/io/fits/tests/test_header.py (hunk 6 above)
Comment (saimn) [inline astropy/io/fits/tests/test_header.py]: Ok, I'll take a look at these tests.

**Comment by saimn** [comment]
@MSeifert04 - Good point for the docs, I forgot to look. It should be better now.

**Comment by pllim** [comment]
> who knows what it will break at STScI

@embray probably everything but it has to happen sooner or later... :joy:

**Comment by pllim** [comment]
Needs a change log.

**Comment by embray** [comment]
> Needs a change log.

:+1: There are examples of this in the past.  "The following previously deprecated interfaces have been removed:
   * ...
   * ...
   * ...
" etc.

**Comment by saimn** [comment]
Added changelog, as well some of the `Header` tests adapted for the new API.

File: astropy/io/fits/header.py [hunk 7]
```diff
@@ -1000,12 +985,9 @@ def update(self, *args, **kwargs):
         .. warning::
             As this method works similarly to `dict.update` it is very
             different from the ``Header.update()`` method in PyFITS versions
```
Comment (MSeifert04) [inline astropy/io/fits/header.py]: Is there a corresponding astropy version? I don't think providing the PyFITS version will help any reader.

File: astropy/io/fits/header.py [hunk 8]
```diff
@@ -1044,132 +1026,48 @@ def update(self, *args, **kwargs):
...
+                raise ValueError(
+                    'Header update value for key %r is invalid; the '
+                    'value must be either a scalar, a 1-tuple '
+                    'containing the scalar value, or a 2-tuple '
+                    'containing the value and a comment string.' % k)
+            self._update(card)
+
+        if other is None:
+            pass
+        elif hasattr(other, 'iteritems'):
```
Comment (MSeifert04) [inline astropy/io/fits/header.py]: This seems like some weird python2-only patch. I don't know if it's out of scope for this PR but I think this should be fixed or at least deserves a `#FIXME` comment.

File: astropy/io/fits/tests/test_image.py [hunk 9]
```diff
@@ -253,7 +248,7 @@ def test_io_manipulation(self):
         hdu2 = fits.ImageHDU(header=r[1].header, data=np.array([1, 2],
                              dtype='int32'))
 
-        assert (str(hdu2.header.ascard[1:5]) ==
+        assert ('\n'.join(str(x) for x in hdu2.header.cards[1:5]) ==
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_image.py]: Maybe that's a stupid question but the old way didn't work anymore?

**Comment by MSeifert04** [comment]
> I think I've always been hesitant to remove them because who knows what it will break at STScI. 

Maybe this should also be announced on the mailing list?

File: astropy/io/fits/header.py (hunk 7 above)
Comment (saimn) [inline astropy/io/fits/header.py]: Indeed, I think it corresponds to Astropy 0.4.

File: astropy/io/fits/header.py (hunk 8 above)
Comment (saimn) [inline astropy/io/fits/header.py]: I agree it could be changed but I prefer not to do this in this PR.
And anyway on python 3 it works with the next `elif hasattr(other, 'keys')`.

File: astropy/io/fits/tests/test_image.py (hunk 9 above)
Comment (saimn) [inline astropy/io/fits/tests/test_image.py]: Nope, `ascard` was returning a `CardList` which had directly a string representation. `.cards` now return a list of `Cards`, so one need to get and join the string representation of each card. 
Another option could be `str(hdu2.header[1:5]))` but not fully equivalent as it includes the `END` keyword.

File: astropy/io/fits/header.py (hunk 7 above)
Comment (saimn) [inline astropy/io/fits/header.py]: No actually PyFITS v3.1 == Astropy v0.1

Comment (embray) [inline astropy/io/fits/header.py]: Yup, that sounds about right.

File: astropy/io/fits/tests/test_image.py (hunk 9 above)
Comment (embray) [inline astropy/io/fits/tests/test_image.py]: Hmm, too bad, maybe there was a (narrow) use for `CardList` after all :)

**Comment by saimn** [comment]
> @embray : who knows what it will break at STScI.
> @MSeifert04 : Maybe this should also be announced on the mailing list?

Do you think this code is still widely used ? From a search on Github with characteristic method names such as `ascardimage`, it seems not. Maybe one concern is for code that still use PyFITS and would migrate to Astropy ?

Also the transition to the new interface is easy for most of the methods. To help people, just in case, I have added a link to the "Header Interface Transition Guide" in the changelog.

**Comment by pllim** [comment]
There are some really old codes at STScI, so my bet is this is bound to break something. But it is probably for the best, as it really forces us to update those codes a little...

**Comment by astrofrog** [comment]
@pllim - and if there are some old codes that really do break and can't be updated, they can rely on the Astropy LTS?

**Comment by pllim** [comment]
@astrofrog - I don't see that situation happening yet, but that can be one possibility.

**Comment by embray** [comment]
Even code that still uses PyFITS should have transitioned a long time ago.  These interfaces were deprecated there before they were ever even moved into Astropy.

And yes, really any critical code that's using Astropy should be on an LTS..........




...............

**Comment by nden** [comment]
We've made the transition to astropy.fits a long time ago. If something was missed, I'd say it's about time to find this out.

**Comment by saimn** [comment]
Ok then, let's go ! :rocket:

File: astropy/io/fits/header.py (hunk 8 above)
Comment (astrofrog) [inline astropy/io/fits/header.py]: We should change this to combine the two ``elif``s into a single one that uses ``six.iteritems`` (but I agree you can do it in a separate PR)

**Comment by astrofrog** [comment]
Looks good to me 👍

**Comment by embray** [comment]
This won't cause anything to break anyways until and unless someone purposefully upgrades to v1.3.0 (granted lots of users just "upgrade" without thinking about it but that's not our fault.......)

Someone with contacts to INS ( @SaraOgaz ? @pllim ? ) might want to cast around and see if this will break any astronomers' / RIA's personal scripts that they have floating around and use regularly, just to give them the heads up.  They often write scripts that produce a *lot* of output and treat miscellaneous warnings as noise :)

**Comment by pllim** [comment]
@justincely might have more recent INS connection than me... :sweat_smile:

File: astropy/io/fits/header.py (hunk 8 above)
Comment (saimn) [inline astropy/io/fits/header.py]: @astrofrog - The loop already uses `six.iteritems`, so I think the simplest change would be to use `hasattr(other, 'items')` and indeed rely on six to do he right thing. I can push a PR later with this.

**Comment by justincely** [comment]
@pllim, @nden sent around a compilation of breaking things, which helps me see what's changing.  I imagine there are definitely people in INS and the scistaff that this will break stuff for.  I'll take what she sent around and get something together to distribute to INS

**Comment by pllim** [comment]
No objection to merge from me.

File: astropy/io/fits/hdu/compressed.py (hunk 4 above)
Comment (embray) [inline astropy/io/fits/hdu/compressed.py]: @bsipocz Of that I have no doubt.
